_posts/2020-10-27-scalajs-for-scala-3.md

```diff
@@ -0,0 +1,184 @@
+This is how Scala.js is able to keep up-to-date with all the versions of Scala 2, and be available on the same day as new Scala releases.
+Not all compile-to-JS languages have that property!
+
+### `.sjsir`files
+
+.sjsir files are similar to .class files, although they are AST-based (instead of stack-machine-based) and contain features dedicated to JavaScript interoperability.
+The .sjsir format and specification is independent of Scala, its compiler or its standard library.
+In fact, there is nothing Scala-specific in the entire linker, other than a few optimizations targeted at Scala-like code.
+This is important because it means that the linker is independent of the version of Scala: we use the same linker irrespective of the version of Scala that was used to create the .sjsir files.
+In theory, we could even compile other JVM languages like Kotlin or Java to the Scala.js IR, then use our linker to emit JavaScript code.
+While we have a proof-of-concept for Java, it is nowhere near usable, however.
+
+.sjsir files have binary compatibility guarantees similar to those of .class files.
+With a few exceptions related to JS interop annotations, if changes to a .class files are binary compatible, then the changes applied to the corresponding .sjsir files are also binary compatible.
+This allows Scala.js codebase to meaningfully use MiMa to check that they do not break their binary API.
+
+The Scala.js linker (optimizer) takes the .sjsir files from the application and from all the transitive dependencies.
+In particular, the .sjsir files that are built for the Scala standard library and for the JDK subset.
+The linker optimizes all of them together, then emits one .js file for the entire codebase.
+
+### Components of Scala.js core
+
+The components involved here have the following sizes (not counting test files):
+
+![Sizes of the components of Scala.js](/resources/img/blog/scalajs-for-scala-3/components-sizes.svg)
+
+The two biggest components, "scalac compiler" and "Scala standard library", are not part of Scala.js itself.
+They are reused from the core Scala repository.
+
+Among the components of Scala.js itself, the linker/optimizer and the subset of the JDK are the biggest, and account for 74% of the total.
+In comparison, the 8,000 lines of code of the Scala.js compiler plugin represent only 11% of the total 69 kLoC.
+
+Add to that about 68,000 lines of testing code, and we get close to 140,000 lines of code for [the Scala.js core repository](https://github.com/scala-js/scala-js).
+For fun, we can compare that to the sizes of the repositories for Scala 2 and Scala 3:
+
+![Sizes of the core Scala repositories](/resources/img/blog/scalajs-for-scala-3/repositories-sizes.svg)
+
+It's worth pointing out that Scala 3 reuses the Scala standard library of Scala 2, so basically those two repositories have the same size.
+
+## Supporting Scala.js in Scala 3
+
+Scala 3 has an entirely new compiler codebase, reimplemented from scratch.
```

Comment:
This isn't really true — the back end was copied, not reimplemented.

Reply:
That's a bit pedantic, though, don't you think? Also it has now diverged quite a bit, so it's not even a simple copy-paste anymore. It is almost as much a reimplementation as the Scala.js back-end was.

Reply:
I didn't know it had diverged that far.